src/render_tree.rs

```diff
@@ -0,0 +1,563 @@
+pub mod properties;
+pub mod text;
+pub mod util;
+
+use std::borrow::BorrowMut;
+use std::{cell::RefCell, rc::Rc};
+
+use crate::html5::parser::document;
+use crate::html5::parser::document::{Document, DocumentHandle};
+
+use crate::html5::node::NodeData;
+use crate::render_tree::{properties::Rectangle, text::TextNode};
+
+/// The position of the render cursor used to determine where
+/// to draw an object
+#[derive(Debug, PartialEq)]
+pub struct Position {
+    pub x: f64,
+    pub y: f64,
+}
+
+impl Position {
+    pub fn new() -> Self {
+        Self { x: 0., y: 0. }
+    }
+
+    pub fn new_from_existing(position: &Position) -> Self {
+        Self {
+            x: position.x,
+            y: position.y,
+        }
+    }
+
+    /// Move position to (x, y)
+    pub fn move_to(&mut self, x: f64, y: f64) {
+        self.x = x;
+        self.y = y;
+    }
+
+    /// Move position relative to another position.
+    /// x = relative.x + x_offset
+    /// y = relative.y + y_offset
+    pub fn move_relative_to(&mut self, relative_position: &Position, x_offset: f64, y_offset: f64) {
+        self.x = relative_position.x + x_offset;
+        self.y = relative_position.y + y_offset;
+    }
+
+    /// Adjust y by an offset.
+    /// y += offset_y
+    pub fn offset_y(&mut self, offset_y: f64) {
+        self.y += offset_y;
+    }
+
+    /// Adjust x by an offset.
+    /// x += offset_x
+    pub fn offset_x(&mut self, offset_x: f64) {
+        self.x += offset_x;
+    }
+}
+
+impl Default for Position {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+/// A RenderTree is a data structure to be consumed by a user agent
+/// that combines the DOM and CSSOM to compute layouts and styles
+/// for objects to draw on the screen.
+pub struct RenderTree {
+    /// Pointer to the underlying document that builds this render tree
+    // TODO: Add CSSOM here as well when we get to that point
+    document: DocumentHandle,
+    /// Entry point of the render tree
+    // TODO: make this a NodeHandle to make operations easier and not
+    // have to keep doing borrow(), borrow_mut() etc.
+    pub root: Rc<RefCell<Node>>,
+    /// Global render cursor position managed by the tree
+    /// that determines where to draw objects on the screen.
+    /// Whenever a node is created, the current value of the
+    /// position is copied into the node.
+    pub position: Position,
+}
+
+impl RenderTree {
+    pub fn new(document: &DocumentHandle) -> Self {
+        Self {
+            document: Document::clone(document),
+            root: Rc::new(RefCell::new(Node::new())),
+            position: Position::new(),
+        }
+    }
+
+    pub fn build(&mut self) {
+        // start with a clean root (if build is called multiple times)
+        self.root = Rc::new(RefCell::new(Node::new()));
```

Comment:
something I am not sure of, if we replace the root like this, are all the children of the (previous) tree deallocated? Does this cause a leak? Checking for leaks on MacOS is terrible compared to linux (I miss valgrind...)